src/tasks/processBulkUpload.ts

```diff
@@ -1,6 +1,88 @@
-import { isProcessBulkUploadJobPayload } from '../types';
-import { InternalValidationError } from '../errors';
+import fs from 'fs';
+import { Readable } from 'stream';
+import { finished } from 'stream/promises';
+import fetch from 'node-fetch';
+import tmp from 'tmp';
+import { db } from '../database/db';
+import { loadBulkUpload } from '../database/accessors';
+import {
+  BulkUploadStatus,
+  isProcessBulkUploadJobPayload,
+} from '../types';
+import {
+  InternalValidationError,
+  NotFoundError,
+} from '../errors';
 import type { JobHelpers } from 'graphile-worker';
+import type { FileResult } from 'tmp';
+import type { Response } from 'node-fetch';
+import type { BulkUpload } from '../types';
+
+const updateBulkUploadStatus = async (
+  id: number,
+  status: BulkUploadStatus,
+): Promise<BulkUpload> => {
+  const bulkUploadsQueryResult = await db.sql<BulkUpload>('bulkUploads.updateStatusById', {
+    id,
+    status,
+  });
+  const bulkUpload = bulkUploadsQueryResult.rows[0];
+  if (bulkUpload === undefined) {
+    throw new NotFoundError(`The bulk upload was not found (id: ${id})`);
+  }
+  return bulkUpload;
+};
+
+const createTemporaryFile = async () => (
+  new Promise<FileResult>((resolve, reject) => {
+    tmp.file((err, name, fd, removeCallback) => {
+      if (err) {
+        reject(err);
+        return;
+      }
+      resolve({
+        name,
+        fd,
+        removeCallback,
+      });
+    });
+  })
+);
+
+const downloadUrlToTemporaryStorage = async (url: string): Promise<FileResult> => {
+  let temporaryFile: FileResult;
+  try {
+    temporaryFile = await createTemporaryFile();
+  } catch (err) {
+    throw new Error('Unable to create a temporary file');
+  }
+
+  let sourceResponse: Response;
+  try {
+    sourceResponse = await fetch(url);
```

Comment:
... darn!  (yep.)

Reply:
OK I looked into this!

I think keeping what is written makes sense for *this commit* but I'm going to add another commit which improves the system to support `S3://` URIs for source files, and if that protocol is set it would process the file using our S3 client.  This would also involve changing the API to allow `S3://` strings as valid sourceUrls (technically I think it's a URI not a URL 😬).

Reply:
@slifty and I talked about this on a call, and I think where we landed was that the `POST /bulkUploads` API needs to be updated to accept the `key` value returned in the presigned URL API instead of a `sourceUrl`, and the worker will compose that key with the bucket to craft an S3 URL. That allows us to avoid pushing too much responsibility to the client, and prevent the potential for mistakes/attacks around client-provided bucket names. It'd also allow us to move from one bucket to another more gracefully, should the need arise.

We then discussed if it's worth holding on to the generic HTTP download code that's already written, and unfortunately decided not to; it'd at least be dead code for the foreseeable future, and maybe also an attack surface. If and when we have partners that want to use this but want us to fetch their CSVs rather than upload them to our S3 bucket, we'll figure out how we want to handle that.

Reply:
The tests will live on in our memories!

(and also the S3 versions of the tests are very similar anyway 😅)

Reply:
This is now sorted / refactored!  In addition to being correct, the dependency tree is all the cleaner for it.